File: backend/pkg/core/datasource/recommendation_service.go
package datasource

import (
	"context"
	"net/url"
	"sort"
	"strings"

	"backend/pkg/api/dto"
	"backend/pkg/core"
	"backend/pkg/database/repository"
	"backend/pkg/integrations/exa"
	"backend/pkg/types"

	"github.com/google/uuid"
)

const (
	defaultRecommendationLimit = 8
	maxRecommendationLimit     = 12
	maxDiscoverySeedSources    = 5
	maxDiscoveryResultsPerSeed = 6
)

// RecommendationSearchService defines the search provider used for source recommendations.
type RecommendationSearchService interface {
	Search(ctx context.Context, query string, options *exa.SearchOptions) (*exa.SearchResponse, error)
	FindSimilar(ctx context.Context, url string, options *exa.FindSimilarOptions) (*exa.SearchResponse, error)
	IsConfigured() bool
}

// RecommendationService recommends data sources from a freeform query without persisting them.
type RecommendationService struct {
	dataSourceRepo repository.DataSourceRepository
	searchService  RecommendationSearchService
}

// NewRecommendationService creates a recommendation service.
func NewRecommendationService(dataSourceRepo repository.DataSourceRepository, searchService RecommendationSearchService) *RecommendationService {
	return &RecommendationService{
		dataSourceRepo: dataSourceRepo,
		searchService:  searchService,
	}
}

// Recommend returns ephemeral source recommendations for the given owner.
func (s *RecommendationService) Recommend(ctx context.Context, orgID *uuid.UUID, userID *uuid.UUID, req dto.DataSourceRecommendationRequest) (*dto.DataSourceRecommendationsResponse, error) {
	if orgID == nil && userID == nil {
		return nil, core.InvalidInputError("Either organization_id or user_id must be provided")
	}
	if s.searchService == nil || !s.searchService.IsConfigured() {
		return nil, core.ExternalError("Source recommendations are unavailable because Exa is not configured")
	}

	limit := normalizeRecommendationLimit(req.Limit)
	existingSources, err := s.listExistingSources(ctx, orgID, userID)
	if err != nil {
		return nil, err
	}

	searchResp, err := s.searchService.Search(ctx, req.Query, &exa.SearchOptions{
		NumResults:        limit * 3,
		UseAutoprompt:     true,
		IncludeText:       true,
		IncludeHighlights: true,
		IncludeSummary:    true,
	})
	if err != nil {
		return nil, core.ExternalError("Failed to fetch source recommendations")
	}

	recommendations := s.buildRecommendations(searchResp.Results, existingSources, limit)
	return &dto.DataSourceRecommendationsResponse{
		Mode:            "query",
		Query:           strings.TrimSpace(req.Query),
		Recommendations: recommendations,
	}, nil
}

// RecommendFromExistingSources returns ephemeral recommendations based on current manual sources.
func (s *RecommendationService) RecommendFromExistingSources(ctx context.Context, orgID *uuid.UUID, userID *uuid.UUID, req dto.DataSourceDiscoveryRecommendationRequest) (*dto.DataSourceRecommendationsResponse, error) {
	if orgID == nil && userID == nil {
		return nil, core.InvalidInputError("Either organization_id or user_id must be provided")
	}
	if s.searchService == nil || !s.searchService.IsConfigured() {
		return nil, core.ExternalError("Source discovery is unavailable because Exa is not configured")
	}

	limit := normalizeRecommendationLimit(req.Limit)
	existingSources, err := s.listExistingSources(ctx, orgID, userID)
	if err != nil {
		return nil, err
	}

	seedSources := filterDiscoverySeedSources(existingSources)
	if len(seedSources) == 0 {
		return &dto.DataSourceRecommendationsResponse{
			Mode:            "discovery",
			Recommendations: []dto.DataSourceRecommendationResponse{},
		}, nil
	}
	if len(seedSources) > maxDiscoverySeedSources {
		seedSources = seedSources[:maxDiscoverySeedSources]
	}

	candidates := make([]discoveryCandidate, 0, len(seedSources)*maxDiscoveryResultsPerSeed)
	failedLookups := 0
	for _, seed := range seedSources {
		resp, findErr := s.searchService.FindSimilar(ctx, seed.URL, &exa.FindSimilarOptions{
			NumResults:          maxDiscoveryResultsPerSeed,
			ExcludeSourceDomain: true,
			IncludeHighlights:   true,
			IncludeSummary:      true,
			IncludeText:         true,
		})
		if findErr != nil {
			failedLookups++
			continue
		}

		for _, result := range resp.Results {
			candidates = append(candidates, discoveryCandidate{
				Result:   result,
				SeedName: seed.Name,
			})
		}
	}

	if len(candidates) == 0 && failedLookups > 0 {
		return nil, core.ExternalError("Failed to fetch adjacent source recommendations")
	}

	sort.SliceStable(candidates, func(i, j int) bool {
		return candidates[i].Result.Score > candidates[j].Result.Score
	})

	recommendations := s.buildDiscoveryRecommendations(candidates, existingSources, seedSources, limit)
	return &dto.DataSourceRecommendationsResponse{
		Mode:            "discovery",
		SeedCount:       len(seedSources),
		Recommendations: recommendations,
	}, nil
}

func (s *RecommendationService) listExistingSources(ctx context.Context, orgID *uuid.UUID, userID *uuid.UUID) ([]types.DataSource, error) {
	if orgID != nil {
		return s.dataSourceRepo.FindByOrganizationID(ctx, *orgID)
	}
	return s.dataSourceRepo.FindByUserID(ctx, *userID)
}

func (s *RecommendationService) buildRecommendations(results []exa.SearchResult, existingSources []types.DataSource, limit int) []dto.DataSourceRecommendationResponse {
	existingURLs := make(map[string]struct{}, len(existingSources))
	existingDomains := make(map[string]struct{}, len(existingSources))
	for _, source := range existingSources {
		normalized, domain := normalizeRecommendationURL(source.URL)
		if normalized != "" {
			existingURLs[normalized] = struct{}{}
		}
		if domain != "" {
			existingDomains[domain] = struct{}{}
		}
	}

	recommendations := make([]dto.DataSourceRecommendationResponse, 0, limit)
	seenDomains := make(map[string]struct{}, limit)

	for _, result := range results {
		normalizedURL, domain := normalizeRecommendationURL(result.URL)
		if normalizedURL == "" || domain == "" {
			continue
		}

		if _, exists := existingURLs[normalizedURL]; exists {
			continue
		}
		if _, exists := existingDomains[domain]; exists {
			continue
		}
		if _, exists := seenDomains[domain]; exists {
			continue
		}

		recommendations = append(recommendations, dto.DataSourceRecommendationResponse{
			Name:        humanizeDomain(domain),
			URL:         normalizedURL,
			Domain:      domain,
			Summary:     summarizeResult(result),
			Reason:      reasonFromResult(result),
			SourceType:  inferRecommendationSourceType(result),
			Score:       result.Score,
			Favicon:     result.Favicon,
			SampleURL:   result.URL,
			SampleTitle: result.Title,
		})
		seenDomains[domain] = struct{}{}

		if len(recommendations) >= limit {
			break
		}
	}

	return recommendations
}

type discoveryCandidate struct {
	Result   exa.SearchResult
	SeedName string
}

func (s *RecommendationService) buildDiscoveryRecommendations(candidates []discoveryCandidate, existingSources []types.DataSource, seedSources []types.DataSource, limit int) []dto.DataSourceRecommendationResponse {
	existingURLs := make(map[string]struct{}, len(existingSources))
	existingRoots := make(map[string]struct{}, len(existingSources))
	for _, source := range existingSources {
		normalized, domain := normalizeRecommendationURL(source.URL)
		if normalized != "" {
			existingURLs[normalized] = struct{}{}
		}
		if root := extractRecommendationDomainRoot(domain); root != "" {
			existingRoots[root] = struct{}{}
		}
	}

	seedRoots := make(map[string]struct{}, len(seedSources))
	for _, source := range seedSources {
		_, domain := normalizeRecommendationURL(source.URL)
		if root := extractRecommendationDomainRoot(domain); root != "" {
			seedRoots[root] = struct{}{}
		}
	}

	recommendations := make([]dto.DataSourceRecommendationResponse, 0, limit)
	seenRoots := make(map[string]struct{}, limit)

	for _, candidate := range candidates {
		normalizedURL, domain := normalizeRecommendationURL(candidate.Result.URL)
		if normalizedURL == "" || domain == "" {
			continue
		}

		root := extractRecommendationDomainRoot(domain)
		if root == "" {
			continue
		}

		if _, exists := existingURLs[normalizedURL]; exists {
			continue
		}
		if _, exists := existingRoots[root]; exists {
			continue
		}
		if _, exists := seedRoots[root]; exists {
			continue
		}
		if _, exists := seenRoots[root]; exists {
			continue
		}

		recommendations = append(recommendations, dto.DataSourceRecommendationResponse{
			Name:        humanizeDomain(domain),
			URL:         normalizedURL,
			Domain:      domain,
			Summary:     summarizeResult(candidate.Result),
			Reason:      discoveryReasonFromCandidate(candidate),
			SourceType:  inferRecommendationSourceType(candidate.Result),
			Score:       candidate.Result.Score,
			Favicon:     candidate.Result.Favicon,
			SampleURL:   candidate.Result.URL,
			SampleTitle: candidate.Result.Title,
		})
		seenRoots[root] = struct{}{}

		if len(recommendations) >= limit {
			break
		}
	}

	return recommendations
}

func normalizeRecommendationLimit(limit int) int {
	switch {
	case limit <= 0:
		return defaultRecommendationLimit
	case limit > maxRecommendationLimit:
		return maxRecommendationLimit
	default:
		return limit
	}
}

func normalizeRecommendationURL(rawURL string) (string, string) {
	parsed, err := url.Parse(strings.TrimSpace(rawURL))
	if err != nil || parsed.Scheme == "" || parsed.Host == "" {
		return "", ""
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return "", ""
	}

	host := strings.ToLower(parsed.Hostname())
	if host == "" {
		return "", ""
	}

	normalized := parsed.Scheme + "://" + host
	return normalized, host
}

func summarizeResult(result exa.SearchResult) string {
	if summary := strings.TrimSpace(result.Summary); summary != "" {
		return summary
	}
	if len(result.Highlights) > 0 {
		return strings.TrimSpace(result.Highlights[0])
	}
	if text := strings.TrimSpace(result.Text); text != "" {
		if len(text) > 220 {
			return strings.TrimSpace(text[:220]) + "..."
		}
		return text
	}
	return ""
}

func reasonFromResult(result exa.SearchResult) string {
	parts := make([]string, 0, 3)
	if result.Title != "" {
		parts = append(parts, result.Title)
	}
	if len(result.Highlights) > 0 {
		parts = append(parts, strings.TrimSpace(result.Highlights[0]))
	}
	if len(parts) == 0 {
		return "Relevant source from AI search results"
	}
	return strings.Join(parts, " - ")
}

func inferRecommendationSourceType(result exa.SearchResult) string {
	combined := strings.ToLower(strings.Join([]string{result.URL, result.Title, result.Summary}, " "))

	switch {
	case strings.Contains(combined, "substack"), strings.Contains(combined, "newsletter"):
		return "newsletter"
	case strings.Contains(combined, "/feed"), strings.Contains(combined, " rss"), strings.Contains(combined, "rss "):
		return "rss"
	case strings.Contains(combined, "forum"), strings.Contains(combined, "community"), strings.Contains(combined, "discuss"), strings.Contains(combined, "reddit"), strings.Contains(combined, "news.ycombinator.com"):
		return "forum"
	case strings.Contains(combined, "news"), strings.Contains(combined, "press"):
		return "news"
	default:
		return "blog"
	}
}

func humanizeDomain(domain string) string {
	base := strings.TrimPrefix(domain, "www.")
	parts := strings.Split(base, ".")
	if len(parts) == 0 || parts[0] == "" {
		return domain
	}

	nameParts := strings.FieldsFunc(parts[0], func(r rune) bool {
		return r == '-' || r == '_'
	})
	for i, part := range nameParts {
		if part == "" {
			continue
		}
		nameParts[i] = strings.ToUpper(part[:1]) + part[1:]
	}

	if len(nameParts) == 0 {
		return domain
	}
	return strings.Join(nameParts, " ")
}

func filterDiscoverySeedSources(sources []types.DataSource) []types.DataSource {
	seeds := make([]types.DataSource, 0, len(sources))
	for _, source := range sources {
		if source.IsEnabled && !source.IsDiscovered {
			seeds = append(seeds, source)
		}
	}
	return seeds
}

func extractRecommendationDomainRoot(domain string) string {
	host := strings.TrimPrefix(strings.ToLower(strings.TrimSpace(domain)), "www.")
	if host == "" {
		return ""
	}

	parts := strings.Split(host, ".")
	if len(parts) < 2 {
		return host
	}

	return parts[len(parts)-2] + "." + parts[len(parts)-1]
}

func discoveryReasonFromCandidate(candidate discoveryCandidate) string {
	if candidate.SeedName == "" {
		return "Adjacent site related to your current sources"
	}
	if title := strings.TrimSpace(candidate.Result.Title); title != "" {
		return "Similar to " + candidate.SeedName + " - " + title
	}
	return "Similar to " + candidate.SeedName
}

File: backend/src/core/datasource/recommendation_service.rs
use std::{collections::BTreeSet, sync::Arc};

use reqwest::Url;
use uuid::Uuid;

use crate::error::AppError;

use super::{
    DataSource, DataSourceDiscoveryRecommendationRequest, DataSourceRecommendationRequest,
    DataSourceRecommendationResponse, DataSourceRecommendationsResponse, DataSourceRepository,
    RecommendationSearchPort, SearchOptions, SearchResult, SimilarOptions,
};

const DEFAULT_LIMIT: i32 = 8;
const MAX_LIMIT: i32 = 12;
const MAX_SEEDS: usize = 5;
const RESULTS_PER_SEED: i32 = 6;

#[derive(Clone)]
pub struct RecommendationService {
    data_sources: Arc<dyn DataSourceRepository>,
    search: Arc<dyn RecommendationSearchPort>,
}

impl RecommendationService {
    pub fn new(
        data_sources: Arc<dyn DataSourceRepository>,
        search: Arc<dyn RecommendationSearchPort>,
    ) -> Self {
        Self {
            data_sources,
            search,
        }
    }

    pub async fn recommend(
        &self,
        organization_id: Option<Uuid>,
        user_id: Option<Uuid>,
        request: DataSourceRecommendationRequest,
    ) -> Result<DataSourceRecommendationsResponse, AppError> {
        require_owner(organization_id, user_id)?;
        if !self.search.is_configured() {
            return Err(AppError::External);
        }
        let limit = normalize_limit(request.limit);
        let existing = self.existing(organization_id, user_id).await?;
        let response = self
            .search
            .search(
                &request.query,
                SearchOptions {
                    num_results: limit * 3,
                    use_autoprompt: true,
                    include_text: true,
                    include_highlights: true,
                    include_summary: true,
                    ..SearchOptions::default()
                },
            )
            .await
            .map_err(|_| AppError::External)?;
        Ok(DataSourceRecommendationsResponse {
            mode: "query".to_owned(),
            query: request.query.trim().to_owned(),
            seed_count: 0,
            recommendations: build_recommendations(response.results, &existing, limit as usize),
        })
    }

    pub async fn recommend_from_existing_sources(
        &self,
        organization_id: Option<Uuid>,
        user_id: Option<Uuid>,
        request: DataSourceDiscoveryRecommendationRequest,
    ) -> Result<DataSourceRecommendationsResponse, AppError> {
        require_owner(organization_id, user_id)?;
        if !self.search.is_configured() {
            return Err(AppError::External);
        }
        let existing = self.existing(organization_id, user_id).await?;
        let seeds: Vec<_> = existing
            .iter()
            .filter(|source| source.is_enabled && !source.is_discovered)
            .take(MAX_SEEDS)
            .cloned()
            .collect();
        if seeds.is_empty() {
            return Ok(DataSourceRecommendationsResponse {
                mode: "discovery".to_owned(),
                query: String::new(),
                seed_count: 0,
                recommendations: Vec::new(),
            });
        }

        let mut candidates = Vec::new();
        let mut failures = 0;
        for seed in &seeds {
            match self
                .search
                .find_similar(
                    &seed.url,
                    SimilarOptions {
                        num_results: RESULTS_PER_SEED,
                        exclude_source_domain: true,
                        include_text: true,
                        include_highlights: true,
                        include_summary: true,
                        ..SimilarOptions::default()
                    },
                )
                .await
            {
                Ok(response) => candidates.extend(
                    response
                        .results
                        .into_iter()
                        .map(|result| (result, seed.name.clone())),
                ),
                Err(_) => failures += 1,
            }
        }
        if candidates.is_empty() && failures > 0 {
            return Err(AppError::External);
        }
        candidates.sort_by(|left, right| {
            right
                .0
                .score
                .partial_cmp(&left.0.score)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        Ok(DataSourceRecommendationsResponse {
            mode: "discovery".to_owned(),
            query: String::new(),
            seed_count: seeds.len() as i32,
            recommendations: build_discovery_recommendations(
                candidates,
                &existing,
                &seeds,
                normalize_limit(request.limit) as usize,
            ),
        })
    }

    async fn existing(
        &self,
        organization_id: Option<Uuid>,
        user_id: Option<Uuid>,
    ) -> Result<Vec<DataSource>, AppError> {
        if let Some(id) = organization_id {
            self.data_sources.find_by_organization_id(id).await
        } else if let Some(id) = user_id {
            self.data_sources.find_by_user_id(id).await
        } else {
            Err(AppError::InvalidInput(
                "Either organization_id or user_id must be provided".to_owned(),
            ))
        }
    }
}

fn build_recommendations(
    results: Vec<SearchResult>,
    existing: &[DataSource],
    limit: usize,
) -> Vec<DataSourceRecommendationResponse> {
    let existing_urls: BTreeSet<_> = existing
        .iter()
        .filter_map(|source| normalize_url(&source.url).map(|pair| pair.0))
        .collect();
    let existing_domains: BTreeSet<_> = existing
        .iter()
        .filter_map(|source| normalize_url(&source.url).map(|pair| pair.1))
        .collect();
    let mut seen = BTreeSet::new();
    results
        .into_iter()
        .filter_map(|result| {
            let (url, domain) = normalize_url(&result.url)?;
            if existing_urls.contains(&url)
                || existing_domains.contains(&domain)
                || !seen.insert(domain.clone())
            {
                return None;
            }
            Some(recommendation(result, url, domain, None))
        })
        .take(limit)
        .collect()
}

fn build_discovery_recommendations(
    candidates: Vec<(SearchResult, String)>,
    existing: &[DataSource],
    seeds: &[DataSource],
    limit: usize,
) -> Vec<DataSourceRecommendationResponse> {
    let existing_urls: BTreeSet<_> = existing
        .iter()
        .filter_map(|source| normalize_url(&source.url).map(|pair| pair.0))
        .collect();
    let existing_roots: BTreeSet<_> = existing
        .iter()
        .filter_map(|source| normalize_url(&source.url).map(|pair| domain_root(&pair.1)))
        .collect();
    let seed_roots: BTreeSet<_> = seeds
        .iter()
        .filter_map(|source| normalize_url(&source.url).map(|pair| domain_root(&pair.1)))
        .collect();
    let mut seen = BTreeSet::new();
    candidates
        .into_iter()
        .filter_map(|(result, seed)| {
            let (url, domain) = normalize_url(&result.url)?;
            let root = domain_root(&domain);
            if existing_urls.contains(&url)
                || existing_roots.contains(&root)
                || seed_roots.contains(&root)
                || !seen.insert(root)
            {
                return None;
            }
            Some(recommendation(result, url, domain, Some(seed)))
        })
        .take(limit)
        .collect()
}

fn recommendation(
    result: SearchResult,
    url: String,
    domain: String,
    seed: Option<String>,
) -> DataSourceRecommendationResponse {
    let reason = match seed {
        Some(seed) if seed.is_empty() => "Adjacent site related to your current sources".to_owned(),
        Some(seed) if result.title.trim().is_empty() => format!("Similar to {seed}"),
        Some(seed) => format!("Similar to {seed} - {}", result.title.trim()),
        None if result.title.is_empty() && result.highlights.is_empty() => {
            "Relevant source from AI search results".to_owned()
        }
        None => [
            Some(result.title.as_str()),
            result.highlights.first().map(String::as_str),
        ]
        .into_iter()
        .flatten()
        .map(str::trim)
        .filter(|part| !part.is_empty())
        .collect::<Vec<_>>()
        .join(" - "),
    };
    DataSourceRecommendationResponse {
        name: humanize_domain(&domain),
        url,
        domain,
        summary: summarize(&result),
        reason,
        source_type: infer_source_type(&result),
        score: result.score,
        favicon: result.favicon.clone(),
        sample_url: result.url.clone(),
        sample_title: result.title.clone(),
    }
}

fn normalize_limit(limit: i32) -> i32 {
    if limit <= 0 {
        DEFAULT_LIMIT
    } else {
        limit.min(MAX_LIMIT)
    }
}

fn normalize_url(raw: &str) -> Option<(String, String)> {
    let parsed = Url::parse(raw.trim()).ok()?;
    if !matches!(parsed.scheme(), "http" | "https") {
        return None;
    }
    let host = parsed.host_str()?.to_lowercase();
    Some((format!("{}://{host}", parsed.scheme()), host))
}

fn summarize(result: &SearchResult) -> String {
    if !result.summary.trim().is_empty() {
        return result.summary.trim().to_owned();
    }
    if let Some(value) = result.highlights.first() {
        return value.trim().to_owned();
    }
    let text = result.text.trim();
    if text.len() <= 220 {
        return text.to_owned();
    }
    format!(
        "{}...",
        String::from_utf8_lossy(&text.as_bytes()[..220]).trim()
    )
}

fn infer_source_type(result: &SearchResult) -> String {
    let combined = format!("{} {} {}", result.url, result.title, result.summary).to_lowercase();
    if combined.contains("substack") || combined.contains("newsletter") {
        "newsletter"
    } else if combined.contains("/feed") || combined.contains(" rss") || combined.contains("rss ") {
        "rss"
    } else if [
        "forum",
        "community",
        "discuss",
        "reddit",
        "news.ycombinator.com",
    ]
    .iter()
    .any(|value| combined.contains(value))
    {
        "forum"
    } else if combined.contains("news") || combined.contains("press") {
        "news"
    } else {
        "blog"
    }
    .to_owned()
}

fn humanize_domain(domain: &str) -> String {
    domain
        .trim_start_matches("www.")
        .split('.')
        .next()
        .unwrap_or(domain)
        .split(['-', '_'])
        .filter(|part| !part.is_empty())
        .map(|part| {
            let mut chars = part.chars();
            match chars.next() {
                Some(first) => first.to_uppercase().chain(chars).collect(),
                None => String::new(),
            }
        })
        .collect::<Vec<String>>()
        .join(" ")
}

fn domain_root(domain: &str) -> String {
    let parts: Vec<_> = domain
        .trim()
        .to_lowercase()
        .trim_start_matches("www.")
        .split('.')
        .map(str::to_owned)
        .collect();
    if parts.len() < 2 {
        parts.first().cloned().unwrap_or_default()
    } else {
        format!("{}.{}", parts[parts.len() - 2], parts[parts.len() - 1])
    }
}

fn require_owner(organization_id: Option<Uuid>, user_id: Option<Uuid>) -> Result<(), AppError> {
    if organization_id.is_none() && user_id.is_none() {
        Err(AppError::InvalidInput(
            "Either organization_id or user_id must be provided".to_owned(),
        ))
    } else {
        Ok(())
    }
}
